File: WEEK6/python/bj1_2042.py
import sys

sys.setrecursionlimit(10**6)
input = sys.stdin.readline


def init(node, start, end):
    if start == end:
        tree[node] = arr[start]
        return tree[node]
    mid = (start + end) // 2
    tree[node] = init(node * 2, start, mid) + init(node * 2 + 1, mid + 1, end)
    return tree[node]


def update(node, start, end, index, diff):
    if index < start or index > end:
        return
    tree[node] += diff
    if start != end:
        mid = (start + end) // 2
        update(node * 2, start, mid, index, diff)
        update(node * 2 + 1, mid + 1, end, index, diff)


def sum_tree(node, start, end, left, right):
    if left > end or right < start:
        return 0
    if left <= start and end <= right:
        return tree[node]
    mid = (start + end) // 2
    return sum_tree(node * 2, start, mid, left, right) + sum_tree(
        node * 2 + 1, mid + 1, end, left, right
    )


n, m, k = map(int, input().split())
arr = [0] * (n + 1)
tree = [0] * (n * 4)

for i in range(1, n + 1):
    arr[i] = int(input())

init(1, 1, n)

for _ in range(m + k):
    a, b, c = map(int, input().split())
    if a == 1:
        diff = c - arr[b]
        arr[b] = c
        update(1, 1, n, b, diff)
    else:
        print(sum_tree(1, 1, n, b, c))

File: WEEK6/Js/bj1_2042.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [N, M, K] = input[0].split(' ').map(Number);
const arr = new BigInt64Array(N + 1);
const tree = new BigInt64Array(N * 4);

for (let i = 1; i <= N; i++) {
    arr[i] = BigInt(input[i]);
}

function init(node, start, end) {
    if (start === end) {
        return tree[node] = arr[start];
    }
    const mid = Math.floor((start + end) / 2);
    return tree[node] = init(node * 2, start, mid) + init(node * 2 + 1, mid + 1, end);
}

function update(node, start, end, index, diff) {
    if (index < start || index > end) return;
    tree[node] += diff;
    if (start !== end) {
        const mid = Math.floor((start + end) / 2);
        update(node * 2, start, mid, index, diff);
        update(node * 2 + 1, mid + 1, end, index, diff);
    }
}

function sum(node, start, end, left, right) {
    if (left > end || right < start) return 0n;
    if (left <= start && end <= right) return tree[node];
    const mid = Math.floor((start + end) / 2);
    return sum(node * 2, start, mid, left, right) + sum(node * 2 + 1, mid + 1, end, left, right);
}

init(1, 1, N);

const output = [];
for (let i = N + 1; i < input.length; i++) {
    const [a, b, c] = input[i].split(' ');
    const idx = Number(b);

    if (a === '1') {
        const val = BigInt(c);
        const diff = val - arr[idx];
        arr[idx] = val;
        update(1, 1, N, idx, diff);
    } else {
        output.push(sum(1, 1, N, idx, Number(c)));
    }
}

console.log(output.join('\n'));